### ISSUE TITLE
Error during build on gathering facts with Debian based container

### ISSUE THREAD

**Issue by duffn** [ansible/ansible-container#440]
##### ISSUE TYPE
 - Bug Report

##### container.yml 
```
version: "2"
settings:
  # The Conductor container does the heavy lifting, and provides a portable
  # Python runtime for building your target containers. It should be derived
  # from the same distribution as you're building your target containers with.
  conductor_base: ubuntu:xenial

  # The deployment_output_path is mounted to the Conductor container, and the 
  # `run` and `deployment` commands then write generated Ansible playbooks to it.
  # deployment_output_path: ./ansible-deployment

  # When using the k8s or openshift engines, use the following to authorize with the API.
  # Values set here will be passed to the Ansible modules. Any file paths will be mounted
  # to the conductor container, allowing the `run` command to access the API.
  #k8s_auth:
    # path to a K8s config file
    #config_file:
    # name of a context found within the config.json file
    #context:
    # URL for accessing the K8s API
    #host:
    # An API authentication token
    #api_key:
    # Path to a ca cert file
    #ssl_ca_cert:
    # Path to a cert file
    #cert_file:
    # Path to a key file
    #key_file:
    #boolean, indicating if SSL certs should be validated
    #verify_ssl:

  # When using the k8s or openshift engines, use the following to set the namespace.
  # If not set, the project name will be used. For openshift, the namespace maps to a project,
  # and description and display_name are supported.
  #k8s_namespace:
  #  name:
  #  description:
  #  display_name:

services: 
  # Add your containers here, specifying the base image you want to build from
  # For example:
  
  web:
    from: ubuntu:xenial
    roles:
      - myrole  
  
registries: {}
  # Add optional registries used for deployment. For example:
  #  google:
  #    url: https://gcr.io
  #    namespace: my-cool-project-xxxxxx   
```

##### main.yml 
The `main.yml` file is not required for >=0.9.0.0.

##### OS / ENVIRONMENT
```
Ansible Container, version 0.9.1rc0
Darwin, mng-it-l-060.mycompany.com, 15.6.0, Darwin Kernel Version 15.6.0: Fri Feb 17 10:21:18 PST 2017; root:xnu-3248.60.11.4.1~1/RELEASE_X86_64, x86_64
2.7.10 (default, Oct 23 2015, 19:19:21)
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)] /Users/nicholasduffy/.virtualenvs/ansible-container2/bin/python
```
##### SUMMARY
Building a Debian based container fails during gathering facts. This does not happen with a Redhat based container.

##### STEPS TO REPRODUCE
```
ansible-container --debug build
```

##### EXPECTED RESULTS
The build to finish successfully.

##### ACTUAL RESULTS
The build fails during gathering facts.
Output: https://gist.github.com/duffn/6e075ee60bb75554354786e4fca4e6ae

**Comment by RickSnz** [comment]
I'm having a very similar problem, not sure if i should open a new issue. The problem itself its the same, i cant build with non Redhat conductors, but the output its not the same (with the same container.yml):
https://gist.github.com/RickSnz/99f2908fd7b94925aac273927b4a465b

By the way, i've tried the same with python2.7, same result

**Comment by chouseknecht** [comment]
@RickSnz, thank you for trying Ansible Container, and sharing your feedback. 

The build of the conductor engine is done using the [Dockerfile found here](https://github.com/ansible/ansible-container/blob/develop/container/docker/templates/conductor-dockerfile.j2), which only supports `fedora`, `centos`, `debian`, `ubuntu` and `alpine`.

**Comment by chouseknecht** [comment]
@duffn, thanks for sharing the results using `ubuntu:xenial`. I'll attempt to reproduce this afternoon, and see if I can figure out what's going on.

**Comment by j00bar** [comment]
@RickSnz A clarification on what @chouseknecht shared about the Conductor base image: you can use any base image you want, but we only aim to ensure Ansible build dependencies are met for the distributions that Chris enumerated. If you want to bring your own Conductor base, based on something other than those five, you simply need to ensure Ansible's build dependencies are met in the image you use.

**Comment by chouseknecht** [comment]
PR #443 enables turning off gather_facts. 

Here's my `container.yml` after applying #443 

```
version: "2"
settings:
  conductor_base: ubuntu:xenial
services:
  web:
    from: ubuntu:xenial
    command: ['/bin/false']
    environment:
      ANSIBLE_PYTHON_INTERPRETER: /usr/bin/python3
    roles:
    - role: setup
       gather_facts: no
    - role: apache
       gather_facts: yes
```
Using the dictionary or map form of a role, set `gather_facts` to the desired state, and it will be passed into the play that applies the role to the target container.

**Comment by chouseknecht** [comment]
@duffn @j00bar 

OK. I think I figured out what's up with `gather_facts` not working with `ubuntu:xenial`.  The first issue is that `lsb-core` and `lsb-release` need to be installed in the conductor. That's simple enough to fix. 

The second issue is that when you fire up a xenial conductor, and a target based on xenial, and mount /usr from conductor into the target as /_usr, the `/_usr/bin/lsb-release` command throws this error:

```
/# /_usr/bin/lsb_release
bash: /_usr/bin/lsb_release: /usr/bin/python3: bad interpreter: No such file or directory
```
It's looking for `/usr/bin/python3` instead of `/_usr/bin/python3`. 

Not sure of the best way to fix this. The simple thing would be to make a sym link in the target for /usr/bin/python3 -> /_usr/bin/python3.

**Comment by chouseknecht** [comment]
@j00bar

Possible solution...

Add support in `container.yml` for a `pre_roles` directives. `pre_roles` would accept a list of roles to run prior to executing `gather_facts` and any roles found in `roles`. 

For example: 

```
service: 
  web:
     image: ubuntu:xenial
     pre_roles:
       - role: create-python3-symlink
     roles:
       - role: geerlingguy.apache
       - role: remove-python3-symlink
```

**Comment by chouseknecht** [comment]
FWIW, I proved the creating the symlink solves the immediate problem. Built an image called `xenial-source` using the following Dockerfile:

```
FROM ubuntu:xenial
RUN ln -s /_usr/bin/python3 /usr/bin/python3
```

Then I modified my `container.yml` to use the new image:

```
ersion: "2"
settings:
  conductor_base: ubuntu:xenial

services:
  web:
    from: xenial-source:latest
    command: ['/bin/false']
    roles:
    - role: ping
```

Run `ansible-container --debug --devel build` now succeeds.

**Comment by RickSnz** [comment]
@chouseknecht @j00bar  Thanks for the help, but i should have clarified that i was using ubuntu:xenial as the conductor base, same as duffn, I have tried algo debian:jessie with the same result. The error i'm getting is different than duffn's, but in the same circumstances, that's why i posted it here. Anyway, it's probably because of something in my system's configuration, after checking the debug logs in the conductor image build, its showing the classic "unable to execute 'x86_64-linux-gnu-gcc': No such file or directory" error

**Comment by duffn** [comment]
@RickSnz Related to this perhaps? https://github.com/ansible/ansible-container/issues/437

**Comment by chouseknecht** [comment]
@RickSnz

You'll have to either build a Xenial conductor that has 'gcc' installed or follow the 'running from source' guide on our doc site to pick up the change pointed to by #437

Likely you will next run into the issue we're working on here, which requires the code changes found in #445.

Apologies for the inconvenience. We'll get this straight, and get a patch on PyPi as soon as we can.

**Comment by RickSnz** [comment]
@duffn @chouseknecht Thanks, that is definitively the problem. I should have checked the closed issues ;)

**Comment by chouseknecht** [comment]
From the jagbot-3000 (@j00bar): 

1) We do the modification to lsb_release (or delete it, and let Ansible fall back to  /etc/lsb-release) 
2) We have an option to _not_ bring the Python runtime into the target container.
3) If you're using Xenial, you can't bring the Python runtime into the target container.

Unfortunately, they did not make separate PYTHONPATH environment variables for py2 and py3. And until Ansible supports Py3 out of the box, we can't run modules on Xenial with Py3.

**Comment by chouseknecht** [comment]
Fixed by #453
